src/pybind/mgr/volumes/fs/health.py

```diff
@@ -0,0 +1,111 @@
+    updates -- but, this is much more general purpose.
+
+    If you subclass this, your class becomes a singleton.
+    """
+
+    VOLUME_HEALTH_PREFIX = "Module 'volumes'"
+
+    # internal identifier(s) for various health attributes
+    PURGE_THREAD_HEALTH = "purge_thread_health"
+
+    # "keys" for various health updates sent to `ceph status`
+    PURGE_THREAD_HEALTH_KEY = "MODULE_VOLUME_PURGE_THREAD"
+
+    # TODO: only send updates for mods that have changed.
+    def _send_health_update(self):
+        health = {}
+        volume_client = self._shared_state['meta']['volume_client']
+
+        for mod_key, func in self.dispatch_table.items():
+            log.info("get health: {0}".format(mod_key))
+            mod, mod_health = func[2](self)
+            health[mod] = mod_health
+        log.debug("sending health={}".format(health))
+        volume_client.send_health_update(health)
+
+    def _purge_thread_health_init(self):
+        self._shared_state['health'][HealthUpdateSingleton.PURGE_THREAD_HEALTH] = []
+
+    def _purge_thread_update_state(self, **kwargs):
+        identifier = kwargs['identifier']
+        remove = kwargs.get('remove', False)
+        if remove:
+            try:
+                self._shared_state['health'][HealthUpdateSingleton.PURGE_THREAD_HEALTH].remove(identifier)
+            except ValueError:
+                pass
+        elif identifier not in self._shared_state['health'][HealthUpdateSingleton.PURGE_THREAD_HEALTH]:
+            self._shared_state['health'][HealthUpdateSingleton.PURGE_THREAD_HEALTH].append(identifier)
+
+    def _get_purge_thread_health(self):
+        health = {}
+        num_failed = len(self._shared_state['health'][HealthUpdateSingleton.PURGE_THREAD_HEALTH])
+        if num_failed:
+            health = {
+                "severity"  : "warning",
+                "summary"   : "{0} has {1} failed purge thread(s)".format(HealthUpdateSingleton.VOLUME_HEALTH_PREFIX, num_failed),
+                "count"     : num_failed,
+                "detail"    : ["purge thread(s) \"{0}\" encountered exceptions".format(",".join(self._shared_state['health'][HealthUpdateSingleton.PURGE_THREAD_HEALTH]))]
+            }
+        return HealthUpdateSingleton.PURGE_THREAD_HEALTH_KEY, health
+
+    def _init(self, volume_client):
+        self._shared_state['meta']['volume_client'] = volume_client
+        for mod_key, func in self.dispatch_table.items():
+            log.info("init: {0}".format(mod_key))
+            func[0](self)
+        self._shared_state['meta']['inited'] = True
+
+    # simple dispatch table (init, set, get) for various health
+    # updates.
+    dispatch_table = {
+        PURGE_THREAD_HEALTH : [_purge_thread_health_init,
+                              _purge_thread_update_state,
+                              _get_purge_thread_health],
+        }
+            
+    _shared_state = {
+        'meta': {
+            'volume_client' : None,
```

Comment:
Do you need a volume_client object?  Instead you could've a mgr object here, and use that to set_health_check()  in `_send_health_update()`?